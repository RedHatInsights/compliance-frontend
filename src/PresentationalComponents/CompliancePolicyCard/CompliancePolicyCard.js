import React from 'react';
import propTypes from 'prop-types';
import Truncate from 'react-truncate';
import {
    Card,
    CardHeader,
    CardBody,
    CardFooter,
    Text,
    TextContent,
    TextVariants,
    Grid,
    GridItem
} from '@patternfly/react-core';
import {
    ChartDonut,
    ChartThemeColor,
    ChartThemeVariant
} from '@patternfly/react-charts';
import '../../Charts.scss';
import { fixedPercentage } from '../../Utilities/TextHelper';
import PolicyPopover from './PolicyPopover';
import {
    Link
} from 'react-router-dom';

class CompliancePolicyCard extends React.Component {
    policy = this.props.policy;
    state = {
        cardTitle: <Truncate lines={1}>{this.policy.name}&nbsp;<PolicyPopover policy={this.policy} /></Truncate>
    }

    onMouseover = () => {
        this.setState({
            cardTitle: <React.Fragment>{this.policy.name}&nbsp;<PolicyPopover policy={this.policy} /></React.Fragment>
        });
    }

    onMouseout = () => {
        this.setState({
            cardTitle: <Truncate lines={1}>{this.policy.name}&nbsp;<PolicyPopover policy={this.policy} /></Truncate>
        });
    }

    render() {
        const { cardTitle } = this.state;
        const {
            compliantHostCount, totalHostCount, refId, name, id
        } = this.policy;
        let donutValues = [
            { x: 'Compliant', y: compliantHostCount },
            { x: 'Non-compliant', y: totalHostCount - compliantHostCount }
        ];
        const compliancePercentage = fixedPercentage(Math.floor(100 *
            (donutValues[0].y / (donutValues[0].y + donutValues[1].y))));

        return (
            <Card widget-id={refId}>
                <CardHeader>
                    <TextContent>
                        <Text onMouseEnter={this.onMouseover.bind(this)} onMouseLeave={this.onMouseout.bind(this)}
                            style={{ fontWeight: '500' }} component={TextVariants.h2}>
                            { cardTitle }
                        </Text>
                        <Grid>
                            <GridItem style={{ display: 'inline-flex' }} span={12}>
                                <TextContent>
                                    <span style={{ fontSize: '30px', fontWeight: 'bold' }}>
                                        { compliantHostCount }
                                    </span>
                                    <span style={{ fontWeight: '500' }}>
                                        {' '}of{' '}
                                    </span>
                                    <span style={{ fontSize: '30px', fontWeight: 'bold' }}>
                                        { totalHostCount }
                                    </span>
                                </TextContent>
                            </GridItem>
                            <GridItem span={12}>
                                <Text>
                                    Systems meet compliance threshold
                                </Text>
                            </GridItem>
                        </Grid>
                    </TextContent>
                </CardHeader>
                <CardBody>
                    <Grid>
                        <GridItem style={{ textAlign: 'center' }} span={12}>
                            <div className='chart-inline'>
                                <div className='card-chart-container'>
                                    <ChartDonut data={donutValues}
                                        identifier={name.replace(/ /g, '')}
                                        innerRadius={122}
                                        themeColor={ChartThemeColor.blue}
                                        themeVariant={ChartThemeVariant.light}
                                        title={compliancePercentage}
                                        subTitle="Systems above threshold"
                                        height={300}
                                        width={300}
                                    />
                                </div>
                            </div>
                        </GridItem>
                    </Grid>
                </CardBody>
                <CardFooter>
                    <Grid>
                        <GridItem span={6}>
                            <TextContent>
                                <Text component={TextVariants.small} style={{ fontSize: '16px' }} >
                                    <Link to={'/reports/' + id} >
                                        View report
                                    </Link>
                                </Text>
                                <Text component={TextVariants.small} style={{ fontSize: '16px' }} >
                                    <Link to={'/policies/' + id} >
                                        View policy
                                    </Link>
                                </Text>
                            </TextContent>
                        </GridItem>
                    </Grid>
                </CardFooter>
            </Card>
        );
    };
};

CompliancePolicyCard.propTypes = {
    policy: propTypes.object
};

export default CompliancePolicyCard;
