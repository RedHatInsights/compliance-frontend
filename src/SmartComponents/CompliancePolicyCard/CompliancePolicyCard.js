import React from 'react';
import {
    Link
} from 'react-router-dom';
import Truncate from 'react-truncate';
import {
    Card,
    CardBody,
    Text,
    TextContent,
    TextVariants,
    Grid,
    GridItem
} from '@patternfly/react-core';
import {
    ChartDonut,
    ChartLabel,
    ChartTheme
} from '@patternfly/react-charts';
import '../../Charts.scss';

class CompliancePolicyCard extends React.Component {
    constructor(policy) {
        super();
        this.policy = policy.policy;
        this.state = { cardTitle: <Truncate lines={1}>{this.policy.name}</Truncate> };
        this.onMouseover = this.onMouseover.bind(this);
        this.onMouseout = this.onMouseout.bind(this);
    }

    onMouseover() {
        this.setState({ cardTitle: this.policy.name });
    }

    onMouseout() {
        this.setState({ cardTitle: <Truncate lines={1}>{this.policy.name}</Truncate> });
    }

    render() {
        const compliantHostCount = this.policy.compliant_host_count;
        const totalHostCount = this.policy.total_host_count;
        let donutValues = [
            { x: 'Compliant', y: compliantHostCount },
            { x: 'Non-compliant', y: totalHostCount - compliantHostCount }
        ];
        const compliancePercentage = Math.floor(100 *
            (donutValues[0].y / (donutValues[0].y + donutValues[1].y))) + '%';
        const label = (
            <svg
                className="chart-label"
                height={300}
                width={300}
            >
                <ChartLabel
                    style={{ fontSize: 26 }}
                    text={compliancePercentage}
                    textAnchor="middle"
                    verticalAnchor="middle"
                    x={150}
                    y={135}
                />
                <ChartLabel
                    style={{ fill: '#bbb' }}
                    text="Compliant"
                    textAnchor="middle"
                    verticalAnchor="middle"
                    x={150}
                    y={165}
                />
            </svg>
        );

        return (
            <Card>
                <CardBody>
                    <Text style={{ fontWeight: '500', color: '#bbb' }} component={TextVariants.small}>External Policy</Text>
                    <TextContent>
                        <Text onMouseEnter={this.onMouseover.bind(this)} onMouseLeave={this.onMouseout.bind(this)}
                            style={{ fontWeight: '500' }} component={TextVariants.h4}>
                            {this.state.cardTitle}
                        </Text>
                    </TextContent>
                    <TextContent className="chart-title">
                        <Grid>
                            <GridItem style={{ display: 'inline-flex' }} span={12}>
                                <TextContent>
                                    <span style={{ fontSize: '30px', fontWeight: '500' }}>
                                        { compliantHostCount }
                                    </span>
                                    <span style={{ fontWeight: '500', color: '#bbb' }}>
                                        {' '}of{' '}
                                    </span>
                                    <span style={{ fontSize: '30px', fontWeight: '500' }}>
                                        { totalHostCount }
                                    </span>
                                </TextContent>
                            </GridItem>
                            <GridItem span={8}>
                                <Text style={{ fontWeight: '500', color: '#bbb' }} component={TextVariants.small}>
                                    Systems Compliant
                                </Text>
                            </GridItem>
                        </Grid>
                    </TextContent>
                    <hr/>
                    <Grid>
                        <GridItem style={{ textAlign: 'center' }} span={12}>
                            <div className='chart-inline'>
                                <div className='card-chart-container'>
                                    {label}
                                    <ChartDonut data={donutValues}
                                        identifier={this.policy.name.replace(/ /g, '')}
                                        theme={ChartTheme.light.blue}
                                        height={205}
                                        width={205}
                                    />
                                </div>
                            </div>
                        </GridItem>
                    </Grid>
                    <TextContent style={{ textAlign: 'right' }}>
                        <Text component={TextVariants.small}>
                            <Link to={'/policies/' + this.policy.id} >
                                More Details
                            </Link>
                        </Text>
                    </TextContent>
                </CardBody>
            </Card>
        );
    };
};

export default CompliancePolicyCard;
