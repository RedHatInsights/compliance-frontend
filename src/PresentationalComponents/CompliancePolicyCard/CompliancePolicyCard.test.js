import toJson from 'enzyme-to-json';

import CompliancePolicyCard from './CompliancePolicyCard';

const defaultProps = {
    policy: {
        id: 'ID',
        refId: 'REF_ID',
        name: 'Test Policy',
        compliantHostCount: 10,
        complianceThreshold: 90,
        totalHostCount: 15,
        businessObjective: {
            title: 'BO'
        }
    }
};

describe('CompliancePolicyCard', () => {
    it('expect to render without error', () => {
        const wrapper = shallow(
            <CompliancePolicyCard { ...defaultProps } />
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });
});

describe('CompliancePolicyCard onMousover/onMouseout', () => {
    it('changes to state to the full policy name on hover and back to truncated', () => {
        const wrapper = shallow(
            <CompliancePolicyCard { ...defaultProps } />
        );
        const instance = wrapper.instance();

        instance.onMouseover();
        expect(wrapper.state()).toMatchSnapshot();

        instance.onMouseout();
        expect(wrapper.state()).toMatchSnapshot();
    });
});

