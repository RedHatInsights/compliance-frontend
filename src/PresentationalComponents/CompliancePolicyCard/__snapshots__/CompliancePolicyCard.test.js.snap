// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompliancePolicyCard expect to render without error 1`] = `
<Card
  widget-id="REF_ID"
>
  <CardHeader>
    <TextContent>
      <Text
        component="h2"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "fontWeight": "500",
          }
        }
      >
        <Truncate
          ellipsis="…"
          lines={1}
          trimWhitespace={false}
          width={0}
        >
          Test Policy
           
          <PolicyPopover
            policy={
              Object {
                "businessObjective": Object {
                  "title": "BO",
                },
                "complianceThreshold": 90,
                "compliantHostCount": 10,
                "id": "ID",
                "name": "Test Policy",
                "refId": "REF_ID",
                "totalHostCount": 15,
              }
            }
          />
        </Truncate>
      </Text>
      <Grid>
        <GridItem
          span={12}
          style={
            Object {
              "display": "inline-flex",
            }
          }
        >
          <TextContent>
            <span
              style={
                Object {
                  "fontSize": "30px",
                  "fontWeight": "bold",
                }
              }
            >
              10
            </span>
            <span
              style={
                Object {
                  "fontWeight": "500",
                }
              }
            >
               
              of
               
            </span>
            <span
              style={
                Object {
                  "fontSize": "30px",
                  "fontWeight": "bold",
                }
              }
            >
              15
            </span>
          </TextContent>
        </GridItem>
        <GridItem
          span={12}
        >
          <Text>
            Systems meet compliance threshold
          </Text>
        </GridItem>
      </Grid>
    </TextContent>
  </CardHeader>
  <CardBody>
    <Grid>
      <GridItem
        span={12}
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        <div
          className="chart-inline"
        >
          <div
            className="card-chart-container"
          >
            <ChartDonut
              data={
                Array [
                  Object {
                    "x": "Compliant",
                    "y": 10,
                  },
                  Object {
                    "x": "Non-compliant",
                    "y": 5,
                  },
                ]
              }
              height={300}
              identifier="TestPolicy"
              innerRadius={122}
              subTitle="Systems above threshold"
              themeColor="blue"
              themeVariant="light"
              title="66%"
              width={300}
            />
          </div>
        </div>
      </GridItem>
    </Grid>
  </CardBody>
  <CardFooter>
    <Grid>
      <GridItem
        span={6}
      >
        <TextContent>
          <Text
            component="small"
            style={
              Object {
                "fontSize": "16px",
              }
            }
          >
            <Link
              to="/reports/ID"
            >
              View report
            </Link>
          </Text>
          <Text
            component="small"
            style={
              Object {
                "fontSize": "16px",
              }
            }
          >
            <Link
              to="/policies/ID"
            >
              View policy
            </Link>
          </Text>
        </TextContent>
      </GridItem>
    </Grid>
  </CardFooter>
</Card>
`;

exports[`CompliancePolicyCard onMousover/onMouseout changes to state to the full policy name on hover and back to truncated 1`] = `
Object {
  "cardTitle": <React.Fragment>
    Test Policy
     
    <PolicyPopover
      policy={
        Object {
          "businessObjective": Object {
            "title": "BO",
          },
          "complianceThreshold": 90,
          "compliantHostCount": 10,
          "id": "ID",
          "name": "Test Policy",
          "refId": "REF_ID",
          "totalHostCount": 15,
        }
      }
    />
  </React.Fragment>,
}
`;

exports[`CompliancePolicyCard onMousover/onMouseout changes to state to the full policy name on hover and back to truncated 2`] = `
Object {
  "cardTitle": <Truncate
    ellipsis="…"
    lines={1}
    trimWhitespace={false}
    width={0}
  >
    Test Policy
     
    <PolicyPopover
      policy={
        Object {
          "businessObjective": Object {
            "title": "BO",
          },
          "complianceThreshold": 90,
          "compliantHostCount": 10,
          "id": "ID",
          "name": "Test Policy",
          "refId": "REF_ID",
          "totalHostCount": 15,
        }
      }
    />
  </Truncate>,
}
`;
